vpr/src/route/route_common.cpp

```diff
@@ -414,57 +412,60 @@ void pathfinder_update_path_cost(t_trace* route_segment_start,
-     * ONE MORE net use this routing node.     */
+void pathfinder_update_single_node_occupancy(int inode, int add_or_sub) {
+    /* Updates pathfinder's occupancy by either adding or removing the
+     * usage of a resource node. */
 
     auto& route_ctx = g_vpr_ctx.mutable_routing();
-    auto& device_ctx = g_vpr_ctx.device();
 
     int occ = route_ctx.rr_node_route_inf[inode].occ() + add_or_sub;
     route_ctx.rr_node_route_inf[inode].set_occ(occ);
     // can't have negative occupancy
     VTR_ASSERT(occ >= 0);
-
-    int capacity = device_ctx.rr_nodes[inode].capacity();
-    if (occ < capacity) {
-        route_ctx.rr_node_route_inf[inode].pres_cost = 1.0;
-    } else {
-        route_ctx.rr_node_route_inf[inode].pres_cost = 1.0 + (occ + 1 - capacity) * pres_fac;
-    }
 }
 
-void pathfinder_update_cost(float pres_fac, float acc_fac) {
-    /* This routine recomputes the pres_cost and acc_cost of each routing        *
-     * resource for the pathfinder algorithm after all nets have been routed.    *
-     * It updates the accumulated cost to by adding in the number of extra       *
-     * signals sharing a resource right now (i.e. after each complete iteration) *
-     * times acc_fac.  It also updates pres_cost, since pres_fac may have        *
-     * changed.  THIS ROUTINE ASSUMES THE OCCUPANCY VALUES IN RR_NODE ARE UP TO  *
-     * DATE.                                                                     */
+void pathfinder_update_acc_cost_and_overuse_info(float acc_fac, OveruseInfo& overuse_info) {
+    /* This routine recomputes the acc_cost (accumulated congestion cost) of each       *
+     * routing resource for the pathfinder algorithm after all nets have been routed.   *
+     * It updates the accumulated cost to by adding in the number of extra signals      *
+     * sharing a resource right now (i.e. after each complete iteration) times acc_fac. *
+     * THIS ROUTINE ASSUMES THE OCCUPANCY VALUES IN RR_NODE ARE UP TO DATE.             */
 
-    int occ, capacity;
+    int overuse;
     auto& device_ctx = g_vpr_ctx.device();
     auto& route_ctx = g_vpr_ctx.mutable_routing();
+    size_t overused_nodes = 0, total_overuse = 0, worst_overuse = 0;
 
     for (size_t inode = 0; inode < device_ctx.rr_nodes.size(); inode++) {
-        occ = route_ctx.rr_node_route_inf[inode].occ();
-        capacity = device_ctx.rr_nodes[inode].capacity();
+        overuse = route_ctx.rr_node_route_inf[inode].occ() - device_ctx.rr_nodes[inode].capacity();
 
-        if (occ > capacity) {
-            route_ctx.rr_node_route_inf[inode].acc_cost += (occ - capacity) * acc_fac;
-            route_ctx.rr_node_route_inf[inode].pres_cost = 1.0 + (occ + 1 - capacity) * pres_fac;
-        }
-
-        /* If occ == capacity, we don't need to increase acc_cost, but a change    *
-         * in pres_fac could have made it necessary to recompute the cost anyway.  */
+        if (overuse > 0) {
```

Comment:
Is there ever a scenario where overuse < 0 could occur (which wasn't a bug)? If not probably good to put a VTR_ASSERT_MSG() to that effect.